proxy/src/telemetry/metrics/histogram.rs

```diff
@@ -0,0 +1,126 @@
+use std::{cmp, fmt, iter, slice};
+use std::num::Wrapping;
+use std::marker::PhantomData;
+
+use super::Counter;
+
+/// A series of latency values and counts.
+#[derive(Debug, Clone)]
+pub struct Histogram<V: Into<u64>> {
+    buckets: Vec<(Bucket, Counter)>,
+
+    /// The total sum of all observed latency values.
+    ///
+    /// Histogram sums always explicitly wrap on overflows rather than
+    /// panicking in debug builds. Prometheus' [`rate()`] and [`irate()`]
+    /// queries handle breaks in monotonicity gracefully (see also
+    /// [`resets()`]), so wrapping is less problematic than panicking in this
+    /// case.
+    ///
+    /// Note, however, that Prometheus actually represents this using 64-bit
+    /// floating-point numbers. The correct semantics are to ensure the sum
+    /// always gets reset to zero after Prometheus reads it, before it would
+    /// ever overflow a 52-bit `f64` mantissa.
+    ///
+    /// [`rate()`]: https://prometheus.io/docs/prometheus/latest/querying/functions/#rate()
+    /// [`irate()`]: https://prometheus.io/docs/prometheus/latest/querying/functions/#irate()
+    /// [`resets()`]: https://prometheus.io/docs/prometheus/latest/querying/functions/#resets
+    ///
+    // TODO: Implement Prometheus reset semantics correctly, taking into consideration
+    //       that Prometheus represents this as `f64` and so there are only 52 significant
+    //       bits.
+    pub sum: Wrapping<u64>,
+
+    _p: PhantomData<(V)>,
+}
+
+#[derive(Debug, Eq, PartialEq, Copy, Clone, Hash)]
+pub enum Bucket {
+    Le(u64),
+    Inf,
```

Comment:
In #775, I just used `u64::MAX` to represent the `+Inf` bucket (and had the `fmt` impl for `Bounds` special-case it to format as the string `"+Inf"`). 

It's up to you which approach is better, but using `u64::MAX` rather than an enum variant means that in `observe()`, rather than saying 
```rust
let ok = match *b {
    Bucket::Le(ceiling) => value <= ceiling,
    Bucket::Inf => true,
};
if ok {
    c.incr();
    return;
}
```

you can just say 

```rust
if value <= *b {
    c.incr();
    return;
}
```

Reply:
Alternatively, you could `impl cmp::Ord<u64> for Bucket` (or `impl cmp::Ord<V> for Bucket<V>` if you take my suggestion to make `Bucket` generic) since you're already implementing `Ord<Bucket>` for it...